cranelift/isle/isle/src/printer.rs

```diff
@@ -16,65 +16,59 @@ where
+    // node.to_doc()
+    //     .render(width, out)
+    //     .map_err(|e| Errors::from_io(e, "failed to print isle"))
 }
 
 impl Printable for Defs {
     fn to_doc(&self) -> String {
-        todo!("Meg TODO") 
+        // Confused what this function is achieving...
+        // let sep = RcDoc::hardline().append(Doc::hardline());
+        // RcDoc::intersperse(self.defs.iter().map(|d| d.to_doc()), sep).append(Doc::hardline()) 
     }
 }
 
-// Is this the proper approach?
 impl Printable for Def {
     fn to_doc(&self) -> String {
+        // Missing match arm error?
         match self {
             Def::Pragma(_) => unimplemented!("pragmas not supported"),
             Def::Type(ref t) => {
-                let mut parts = String::from("type");
-                parts.push_str(&t.name.to_doc());
+                let mut parts = format!("(type {} ", &t.name.to_doc());
                 if t.is_extern {
-                    parts.push_str("extern");
+                    parts.push_str("extern ");
                 }
                 if t.is_nodebug {
-                    parts.push_str("nodebug");
+                    parts.push_str("nodebug ");
                 }
-                parts.push_str(&t.ty.to_doc());
-                // Should a return statement be used here?
+                parts.push_str(&format!("{})", &t.ty.to_doc()));
                 return parts
             }
             Def::Rule(ref r) => {
-                let mut parts = String::from("rule");
+                let mut parts = String::from("(rule ");
                 if let Some(prio) = &r.prio {
-                    // Is this the best way to convert from &i64 to string?
-                    parts.push_str(&format!("{}", prio));
-
+                    parts.push_str(&format!("{} ", &prio.to_string()));
                 }
-                parts.push_str(&r.pattern.to_doc());
-                // This function converts to a map of type il? but how is that printed as a string?
-                parts.extend(&r.iflets.iter().map(|il| il.to_doc()));
-                parts.push_str(&r.expr.to_doc());
-                return parts
-            }
-            Def::Extractor(ref e) => {
-                // sexp(vec![
-                //     RcDoc::text("extractor"),
-                //     sexp(
-                //         Vec::from([e.term.to_doc()])
-                //             .into_iter()
-                //             .chain(e.args.iter().map(|v| v.to_doc())),
-                //     ),
-                //     e.template.to_doc(),
-                // ])
-
-                let mut parts = String::from("extractor");
-                parts.push_str(&e.term.to_doc());
-                // Add something else here in the middle
-                parts.push_str(&e.template.to_doc());
+                parts.push_str(&format!(
+                    "{} {} {})", 
+                    &r.pattern.to_doc(),
+                    // join is not defined for Maps
```

Comment:
Still having trouble adding the Map to the string (most of my errors are relating to this problem)